python/lsst/pipe/base/taskFactory.py

```diff
@@ -0,0 +1,97 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Module definig TaskFactory interface.
+"""
+
+__all__ = ["TaskFactory"]
+
+from abc import ABCMeta, abstractmethod
+
+
+class TaskFactory(metaclass=ABCMeta):
+    """Abstract base class for task factory.
+
+    Task factory is responsible for importing PipelineTask subclasses by
+    name and creating instances of these classes.
+    """
+
+    @abstractmethod
+    def loadTaskClass(self, taskName):
+        """Locate and import PipelineTask class.
+
+        Returns tuple of task class and its full name, `None` is returned
+        for both if loading fails.
+
+        Parameters
+        ----------
+        taskName : `str`
+            Name of the PipelineTask class, interpretation depends entirely on
+            activator, e.g. it may or may not include dots.
```

Comment:
Would it be appropriate to say that the interpretation now is now entirely up to the `TaskFactory` and the contexts in which it is used?  Or are there other places where assumptions are made about the form of these names?

Reply:
At this point there are no other assumption, task names are just passed from Pipeline to task factory. TaskFactory implemented in `ctrl_mpexec` expects either something that can be resolved w.r.t. `PYTHONPATH` (e.g. `module.TaskClass`) or just a task class name (super-inefficient to search). Other factories may use different conventions, e.g. registry can be built which uses abstract task names instead of class names, though that would probably make pipeline definitions non-portable.

Reply:
And I may be confusing myself here, I think that logic may be a bit different - TaskFactory is used to resolve any user-provided name to a full class name and that is passed then to a Pipeline. Though even that is an implementation detail now and is not documented anywhere.

Reply:
I don't know that we need to enforce this, but I can't actually think of a use case for having different `TaskFactory` implementations that might expect different kinds of names, and I think it would be desirable for all activators to handle task names consistently.

I gather the reason you didn't want to move the TaskFactory implementation here was because you didn't quite consider it mature (and DM-17044 backs that up).  Perhaps after it does mature we should just move the implementation here and remove the ABC?

Reply:
Well, I think I actually prefer TaskFactory to be an ABC here, at least it tells clients what minimal interface is needed for the rest of the thing to work. Implementation in ctrl_mpexec does a little bit more that is needed from this interface - it also supports functionality that is needed by `pipetask` script like making a list of available tasks.